### PR TITLE
feat(cdk/testing): support modifiers for clicking on a TestElement

### PR DESCRIPTION
**Pull request by wwjhu**
Adds the ability to provide modifier keys when clicking on a TestElement.

### CODE REVIEW

**Comment by mmalerba** [comment]
@wwjhu apologies, I missed this PR while I was out on vacation. It looks good, though it does need to be rebased now.

**Comment by wwjhu** [comment]
Rebased.

**Comment by wwjhu** [comment]
Given the `rightClick` that landed in https://github.com/angular/components/pull/20400, do you think the modifiers should be added to `rightClick` as well?

**Comment by mmalerba** [comment]
Yeah I think that would make sense, people will expect the method signatures to be the same

File: src/cdk/testing/protractor/protractor-element.ts [hunk 1]
```diff
@@ -202,17 +204,33 @@ export class ProtractorElement implements TestElement {
...
-      .mouseMove(await this.element.getWebElement(), ...offsetArgs)
-      .click(button)
-      .perform();
+    // This is the default behavior we want, so we use an empty array of offsetArgs if
+    // no args remain after popping the modifiers from the args passed to this function.
+    const offsetArgs = (args.length === 2 ?
+      [{x: args[0], y: args[1]}] : []) as [{x: number, y: number}];
+
+    let actions = browser.actions()
+    .mouseMove(await this.element.getWebElement(), ...offsetArgs);
```
Comment (mmalerba) [inline src/cdk/testing/protractor/protractor-element.ts:223]: nit: indentation

File: src/cdk/testing/test-element.ts [hunk 2]
```diff
@@ -76,25 +76,27 @@ export interface TestElement {
    * the element is clicked at a specific location, consider using `click('center')` or
    * `click(x, y)` instead.
    */
-  click(): Promise<void>;
+  click(modifiers?: ModifierKeys): Promise<void>;
 
   /** Click the element at the element's center. */
-  click(location: 'center'): Promise<void>;
+  click(location: 'center', modifiers?: ModifierKeys, ): Promise<void>;
```
Comment (mmalerba) [inline src/cdk/testing/test-element.ts:82]: nit: remove trailing comma + whitespace

File: src/cdk/testing/test-element.ts [hunk 3]
```diff
@@ -76,25 +76,27 @@ export interface TestElement {
...
 
   /**
    * Click the element at the specified coordinates relative to the top-left of the element.
    * @param relativeX Coordinate within the element, along the X-axis at which to click.
    * @param relativeY Coordinate within the element, along the Y-axis at which to click.
+   * @param modifiers Modifier keys held while clicking
    */
-  click(relativeX: number, relativeY: number): Promise<void>;
+  click(relativeX: number, relativeY: number, modifiers?: ModifierKeys ): Promise<void>;
```
Comment (mmalerba) [inline src/cdk/testing/test-element.ts:90]: nit: remove whitespace after `ModifierKeys`

File: src/cdk/testing/testbed/unit-test-element.ts [hunk 4]
```diff
@@ -241,10 +248,9 @@ export class UnitTestElement implements TestElement {
...
 
     this._dispatchPointerEventIfSupported('pointerdown', clientX, clientY, button);
-    dispatchMouseEvent(this.element, 'mousedown', clientX, clientY, button);
+    dispatchMouseEvent(this.element, 'mousedown', clientX, clientY, button, modifiers);
     this._dispatchPointerEventIfSupported('pointerup', clientX, clientY, button);
-    dispatchMouseEvent(this.element, 'mouseup', clientX, clientY, button);
-    dispatchMouseEvent(this.element, name, clientX, clientY, button);
-    await this._stabilize();
+    dispatchMouseEvent(this.element, 'mouseup', clientX, clientY, button, modifiers);
+    dispatchMouseEvent(this.element, name, clientX, clientY, button, modifiers);
```
Comment (mmalerba) [inline src/cdk/testing/testbed/unit-test-element.ts:254]: why was `await this._stabilize()` removed?

Comment (wwjhu) [reply src/cdk/testing/testbed/unit-test-element.ts:254]: `_stabilize` is already called from the public entry points `click` and `rightClick` that use this private helper. It's not necessary to call it more often.

**Review by mmalerba — APPROVED** [review]
LGTM

**Comment by wwjhu** [comment]
Rebased.

File: src/cdk/testing/testbed/unit-test-element.ts (hunk 4 above)
Comment (mmalerba) [reply src/cdk/testing/testbed/unit-test-element.ts:254]: can we add a second `await this._stabilize()` to `click` to maintain the same behavior that it has currently? I tried presubmitting and it seems maybe some people were depending on that quirk. You can leave a TODO to clean it up later, just don't want it to block this

Comment (wwjhu) [reply src/cdk/testing/testbed/unit-test-element.ts:254]: Obviously I can, but that second `await this._stabilize()` was recently added in https://github.com/angular/components/pull/20400 which makes that there cannot be many people, if any at all, relying on it

Comment (mmalerba) [reply src/cdk/testing/testbed/unit-test-element.ts:254]: It's possible the failures I saw were not related to this line. I'll do some more digging and let you know

Comment (mmalerba) [reply src/cdk/testing/testbed/unit-test-element.ts:254]: I confirmed that removing that call to `this._stabilize()` was indeed the culprit. It'll need to be investigated, but shouldn't block this PR. Can you add it back with a comment to investigate why removing it breaks some tests in g3?

Comment (wwjhu) [reply src/cdk/testing/testbed/unit-test-element.ts:254]: Will do

**Comment by wwjhu** [comment]
Rebased.

**Pull request merged**